tests/test_csr.py

```diff
@@ -0,0 +1,164 @@
+    assert signing_request.private_key == ""
+    assert signing_request.csr == ""
+    signing_request.save()
+    saved_private_key = signing_request.private_key
+    assert signing_request.private_key != ""
+    assert signing_request.csr != ""
+    # Additional saves do not overwrite the private key
+    signing_request.save()
+    assert signing_request.private_key == saved_private_key
+
+
+@pytest.mark.django_db
+def test_generate_csr():
+    signing_request = SigningRequest.objects.create(
+        common_name="test.com",
+        country_name="US",
+        organization_name="Test Org",
+        state_or_province_name="Test State",
+        email_address="test@test.com",
+    )
+
+    csr = x509.load_pem_x509_csr(signing_request.csr.encode(), default_backend())
+
+    subject = csr.subject
+    assert subject.get_attributes_for_oid(x509.NameOID.COUNTRY_NAME)[0].value == "US"
+    assert (
+        subject.get_attributes_for_oid(x509.NameOID.COMMON_NAME)[0].value == "test.com"
+    )
+    assert (
+        subject.get_attributes_for_oid(x509.NameOID.ORGANIZATION_NAME)[0].value
+        == "Test Org"
+    )
+    assert (
+        subject.get_attributes_for_oid(x509.NameOID.STATE_OR_PROVINCE_NAME)[0].value
+        == "Test State"
+    )
+    assert (
+        subject.get_attributes_for_oid(x509.NameOID.EMAIL_ADDRESS)[0].value
+        == "test@test.com"
+    )
+
+
+@pytest.mark.django_db
+def test_generate_private_key():
+    # A new signing request should not have a private key
+    signing_request = SigningRequest(common_name="test.com")
+    assert signing_request.private_key == ""
+
+    # Generating a private key should populate the private key field
+    generate_private_key(signing_request)
+    saved_private_key = signing_request.private_key
+    assert signing_request.private_key != ""
+    assert "BEGIN PRIVATE KEY" in signing_request.private_key
+
+    # Additional saves do not overwrite the private key
+    signing_request.save()
+    assert signing_request.private_key == saved_private_key
+
+
+@pytest.mark.django_db
+def test_download_csr_single(admin_client):
+    signing_request = SigningRequest.objects.create(
+        common_name="Test", csr="CSR Content"
+    )
+
+    url = reverse("admin:simple_certmanager_signingrequest_changelist")
+    response = admin_client.post(
+        url, {"action": "download_csr", "_selected_action": [signing_request.pk]}
+    )
+
+    assert isinstance(response, FileResponse)
+    assert response["Content-Type"] == "application/pem-certificate-chain"
+    assert "attachment; filename=" in response["Content-Disposition"]
+
+
+@pytest.mark.django_db
+def test_download_csr_multiple(admin_client):
+    csr_content = open("tests/data/valid_csr.pem").read()
+    signing_request1 = SigningRequest.objects.create(
+        common_name="Test", csr=csr_content
+    )
+    signing_request2 = SigningRequest.objects.create(
+        common_name="Test2", csr=csr_content
+    )
```

Comment:
mm, what's the reason for including a fixed file here instead of just letting the private key + csr generation do its job?

I'd expected something like this instead:

```py
SigningRequest.obejcts.create(common_name="test.com", country_name="NL")
```
and that should be baked into the CSR.

Reply:
I did it this way cause I only wanted to test the end of the process (the download) rather than testing the full process but yeah I think it's good to have a full integration test instead.

Reply:
I wouldn't say we are "testing the full process", but rather we are using the new feature in the test setup. the assertions are all about the zip download and that's good :)